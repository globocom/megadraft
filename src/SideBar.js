import React, {Component} from "react";
import ReactDOM from "react-dom";
import {RichUtils} from "draft-js";


const BLOCK_TYPES = [{
  "name": "h2",
  "style": "header-two"
}, {
  "name": "ol",
  "style": "ordered-list-item"
}, {
  "name": "ul",
  "style": "unordered-list-item"
}, {
  "name": "''",
  "style": "blockquote"
}];

class BlockButton extends Component {
  onMouseDown(e) {
    e.preventDefault();
    this.props.onToggle(this.props.blockType.style);
  }

  render() {
    var style = {
      cursor: "pointer",
      display: "inline-block",
      padding: "5px"
    };
    if (this.props.active) {
      style.color = "blue";
    }
    return (
      <li style={style} onMouseDown={::this.onMouseDown}>
        {this.props.blockType.name}
      </li>
    );
  }
}

class BlockStyles extends Component {
  render() {
    const style = {
      listStyle: "none",
      display: this.props.open? "block": "none",
      position: "absolute",
      padding: 0,
      border: "1px solid #CCC",
      top: "-33px",
      backgroundColor: "#FCFCFC",
      boxShadow: "1px 1px 2px 0px #CCC"
    };
    return (
      <ul style={style}>
        {BLOCK_TYPES.map((item) => {
          const active = item.style === this.props.currentBlockType;
          return (
            <BlockButton
              active={active}
              key={item.style}
              onToggle={::this.props.onToggle}
              blockType={item} />
          );
        })}
      </ul>
    );
  }
}

class PopOverMenu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false
    };
  }

  handleMouseOver() {
    this.setState({
      open: true
    });
  }

  handleMouseOut() {
    this.setState({
      open: false
    });
  }

  render() {
    return (
      <li
        style={{position: "relative"}}
        onMouseOver={::this.handleMouseOver}
        onMouseOut={::this.handleMouseOut}>
        T
        <BlockStyles
          open={this.state.open}
          onToggle={this.props.onToggle}
          currentBlockType={this.props.currentBlockType} />
      </li>
    );
  }
}

function getSelectedBlockElement() {
  // Finds the block parent of the current selection
  // https://github.com/facebook/draft-js/issues/45
  const selection = window.getSelection();
  if (selection.rangeCount === 0) {
    return null;
  }
  var node = selection.getRangeAt(0).startContainer;

  do {
    if (node.getAttribute && node.getAttribute("data-block") == "true") {
      return node;
    }
    node = node.parentNode;
  } while (node != null);

}

export default class SideBar extends Component {
  constructor(props) {
    super(props);
    this.state = {top: 0};
  }

  onToggle(blockType) {
    this.props.onChange(
      RichUtils.toggleBlockType(
        this.props.editorState,
        blockType
      )
    );
  }

  componentDidUpdate() {
    if (this.updatingPosition) {
      clearImmediate(this.updatingPosition);
    }
    this.updatingPosition = null ;
    this.updatingPosition = setImmediate(() => {
      return this.setBarPostion();
    });
  }

  setBarPostion() {
    const container = ReactDOM.findDOMNode(this.refs.container);

    const element = getSelectedBlockElement();

    if (!element) {
      return;
    }

    const top = Math.floor(
      element.getBoundingClientRect().top - 21 -
      (container.getBoundingClientRect().top -
        document.documentElement.clientTop));

    if (this.state.top !== top) {
      this.setState({
        top: top
      });
    }
  }

  render() {
    const style = {
      container: {
        cursor: "pointer",
        position: "relative"
      },
      popover: {
        position: "absolute",
        left: "-160px",
        float: "left",
        width: "160px",
        backgroundColor: "#FAFAFA",
        top: `${this.state.top}px`
      }
    };
    const {editorState} = this.props;
    const selection = editorState.getSelection();
    const currentBlockType = editorState
      .getCurrentContent()
      .getBlockForKey(selection.getStartKey())
      .getType();

    return (
      <div ref="container" style={style.container}>
        <div style={style.popover}>
          <ul style={{listStyle: "none"}}>
            <PopOverMenu
              onToggle={::this.onToggle}
              currentBlockType={currentBlockType} />
          </ul>
        </div>
      </div>
    );
  }
}
