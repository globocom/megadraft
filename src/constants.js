/*
 * Copyright (c) 2019, Globo.com (https://github.com/globocom)
 *
 * License: MIT
 */

export const SIDEBAR_EXPAND = "SIDEBAR_EXPAND";
export const SIDEBAR_SHRINK = "SIDEBAR_SHRINK";
export const SIDEBAR_CLICK_MORE = "SIDEBAR_CLICK_MORE";
export const SIDEBAR_ADD_PLUGIN = "SIDEBAR_ADD_PLUGIN";

export const PLUGINS_MODAL_CLOSE = "PLUGINS_MODAL_CLOSE";
export const PLUGINS_MODAL_ADD_PLUGIN = "PLUGINS_MODAL_ADD_PLUGIN";

export const BLOCK_SWAP_UP = "BLOCK_SWAP_UP";
export const BLOCK_SWAP_DOWN = "BLOCK_SWAP_DOWN";
