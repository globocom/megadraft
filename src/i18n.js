/*
 * Copyright (c) 2018, Globo.com (https://github.com/globocom)
 *
 * License: MIT
 */

const i18nConfig = {
  "en-US": {
    "Type the link and press enter": "Type the link and press enter",
    "Invalid Link": "Invalid Link",
    "Can't show plugin, component {{type}} not found.":
      "Can't show plugin, component {{type}} not found.",
    "Can't show plugin, component not found.":
      "Can't show plugin, component not found.",
    "Block List": "Block List",
    "Something went wrong in component '{{type}}'. {{error}}":
      "Something went wrong in component '{{type}}'. {{error}}",
    "Something went wrong with the component type.":
      "Something went wrong with the component type."
  },
  "pt-BR": {
    "Type the link and press enter": "Digite o link e pressione enter",
    "Invalid Link": "Link inválido",
    "Can't show plugin, component {{type}} not found.":
      "Não é possível exibir o plugin, componente {{type}} não encontrado",
    "Can't show plugin, component not found.":
      "Não é possível exibir o plugin, componente não encontrado",
    "Block List": "Lista de Blocos",
    "Something went wrong in component '{{type}}'. {{error}}":
      "Algo esta errado no componente '{{type}}'. {{error}}",
    "Something went wrong with the component type.":
      "Algo esta errado com o tipo do componente"
  }
};

export const replaceData = (str, data) => {
  const rgx = /{{\s?(\w+)\s?}}/gm;
  let msg = str;
  msg = msg.replace(rgx, (_, key) => data[key]);
  return msg;
};

export default i18nConfig;
