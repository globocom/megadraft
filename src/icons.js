/*
 * Copyright (c) 2016, Globo.com (https://github.com/globocom)
 *
 * License: MIT
 */

import BoldIcon from "./icons/bold";
import ItalicIcon from "./icons/italic";
import ULIcon from "./icons/ul";
import OLIcon from "./icons/ol";
import H2Icon from "./icons/h2";
import BlockQuoteIcon from "./icons/blockquote";
import LinkIcon from "./icons/link";
import CrossIcon from "./icons/cross";
import ImageIcon from "./icons/image";
import VideoIcon from "./icons/video";
import EditIcon from "./icons/edit";
import DeleteIcon from "./icons/delete";
import CropIcon from "./icons/crop";
import MediaBigIcon from "./icons/mediaBig";
import MediaMediumIcon from "./icons/mediaMedium";
import MediaSmallIcon from "./icons/mediaSmall";
import DropdownArrow from "./icons/dropdownArrow";
import ErrorIcon from "./icons/error";
import UnlinkIcon from "./icons/unlink";
import CloseIcon from "./icons/close";
import MoreIcon from "./icons/more";


const icons = {
  BoldIcon, ItalicIcon, ULIcon, OLIcon, H2Icon, BlockQuoteIcon, LinkIcon,
  CrossIcon, ImageIcon, VideoIcon, EditIcon, DeleteIcon, CropIcon, MediaBigIcon,
  MediaMediumIcon, MediaSmallIcon, DropdownArrow, ErrorIcon, UnlinkIcon,
  CloseIcon, MoreIcon
};

export default icons;
