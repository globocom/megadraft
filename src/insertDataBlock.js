/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * Copyright (c) 2016, Globo.com (https://github.com/globocom)
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @providesModule AtomicBlockUtils
 * @typechecks
 * @flow
 */

import Immutable from "immutable";

import {
  BlockMapBuilder,
  ContentBlock,
  EditorState,
  genKey,
  Modifier
} from "draft-js";

const { List, Map } = Immutable;

function insertDataBlock(editorState, data) {
  const contentState = editorState.getCurrentContent();
  const selectionState = editorState.getSelection();

  const afterRemoval = Modifier.removeRange(
    contentState,
    selectionState,
    "backward"
  );

  const targetSelection = afterRemoval.getSelectionAfter();
  const afterSplit = Modifier.splitBlock(afterRemoval, targetSelection);
  const insertionTarget = afterSplit.getSelectionAfter();

  const asAtomicBlock = Modifier.setBlockType(
    afterSplit,
    insertionTarget,
    "atomic"
  );

  const block = new ContentBlock({
    key: genKey(),
    type: "atomic",
    text: "",
    characterList: List(),
    data: new Map(data)
  });

  const fragmentArray = [
    block,
    new ContentBlock({
      key: genKey(),
      type: "unstyled",
      text: "",
      characterList: List()
    })
  ];

  const fragment = BlockMapBuilder.createFromArray(fragmentArray);

  const withAtomicBlock = Modifier.replaceWithFragment(
    asAtomicBlock,
    insertionTarget,
    fragment
  );

  const newContent = withAtomicBlock.merge({
    selectionBefore: selectionState,
    selectionAfter: withAtomicBlock.getSelectionAfter().set("hasFocus", true)
  });

  return EditorState.push(editorState, newContent, "insert-fragment");
}

export default insertDataBlock;
