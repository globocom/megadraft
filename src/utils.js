/*
 * Copyright (c) 2016, Globo.com (https://github.com/globocom)
 * Copyright (c) 2016, Andrew Coelho <info@andrewcoelho.com>
 *
 * License: MIT
 */

import {
  convertToRaw,
  convertFromRaw,
  EditorState,
  getVisibleSelectionRect} from "draft-js";

import defaultDecorator from "./decorators/defaultDecorator";


export function editorStateToJSON(editorState) {
  if (editorState) {
    const content = editorState.getCurrentContent();
    return JSON.stringify(convertToRaw(content), null, 2);
  }
}

export function editorStateFromRaw(rawContent, decorator = defaultDecorator) {
  if (rawContent) {
    const content = convertFromRaw(rawContent);
    return EditorState.createWithContent(content, decorator);
  } else {
    return EditorState.createEmpty(decorator);
  }
}

export function getSelectedBlockElement(range) {
  let node = range.startContainer;
  do {
    const nodeIsDataBlock = node.getAttribute ? node.getAttribute("data-block") : null;
    if (nodeIsDataBlock) {
      return node;
    }
    node = node.parentNode;
  } while (node !== null);
  return null;
}

export function getSelectionCoords(editor, toolbar) {
  const editorBounds = editor.getBoundingClientRect();
  const rangeBounds = getVisibleSelectionRect(window);

  if (!rangeBounds || !toolbar) {
    return null;
  }

  const rangeWidth = rangeBounds.right - rangeBounds.left;

  const toolbarHeight = toolbar.offsetHeight;
  // const rangeHeight = rangeBounds.bottom - rangeBounds.top;
  const offsetLeft = (rangeBounds.left - editorBounds.left)
            + (rangeWidth / 2);
  const offsetTop = rangeBounds.top - editorBounds.top - (toolbarHeight + 14);
  const offsetBottom = editorBounds.bottom - rangeBounds.top + 14;
  return {offsetLeft, offsetTop, offsetBottom};
}

export function createTypeStrategy(type) {
  return (contentBlock, callback, contentState) => {
    contentBlock.findEntityRanges(
      (character) => {
        const entityKey = character.getEntity();
        return (
          entityKey !== null &&
          contentState.getEntity(entityKey).getType() === type
        );
      },
      callback
    );
  };
}

/**
 * Returns a wrapper for the given function which cannot be called
 * more often than the given interval. Every time the wrapper is called
 * a timeout gets reset to the interval's number of ms before calling the fn.
 *
 * Keep attention to bind the correct context to the provided funtion using bind() or '::'!
 *
 * @export
 * @param {function} fn The function to execute after the given interval.
 * @param {number} [interval=100] The interval to wait for before calling the wrapped function.
 * @example
 * ```
 * const delayedLog = delayCall(::console.log, 200);
 * delayedLog('hans');
 * delayedLog('heiri');
 * // logs 'heiri' after 200ms, 'hans' won't be logged at all.
 * ```
 * @returns {void}
 */
export function delayCall(fn, interval = 100) {
  let timeout;
  return function (...args) {
    if (timeout) {
      window.clearTimeout(timeout);
    }
    timeout = window.setTimeout(() => fn.apply(window, args), interval);
  };
}
