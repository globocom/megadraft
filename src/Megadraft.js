/*
 * Copyright (c) 2016, Globo.com (https://github.com/globocom)
 *
 * License: MIT
 */

import React, {Component} from "react";
import Draft, {Editor, RichUtils} from "draft-js";

import icons from "./icons";
import Toolbar from "./components/Toolbar";
import Sidebar from "./components/Sidebar";
import Media from "./components/Media";
import DEFAULT_PLUGINS from "./plugins/default";


export default class Megadraft extends Component {
  constructor(props) {
    super(props);

    this.state = {
      readOnly: false
    };

    this.setReadOnly = ::this.setReadOnly;

    this.actions = [
      {type: "inline", label: "B", style: "BOLD", icon: icons.BoldIcon},
      {type: "inline", label: "I", style: "ITALIC", icon: icons.ItalicIcon},
      {type: "entity", label: "Link", style: "link", icon: icons.LinkIcon},
      {type: "separator"},
      {type: "block", label: "UL", style: "unordered-list-item", icon: icons.ULIcon},
      {type: "block", label: "OL", style: "ordered-list-item", icon: icons.OLIcon},
      {type: "block", label: "H2", style: "header-two", icon: icons.H2Icon},
      {type: "block", label: "QT", style: "blockquote", icon: icons.BlockQuoteIcon}
    ];

    this.plugins = this.props.plugins || DEFAULT_PLUGINS;
  }

  onChange(editorState) {
    this.props.onChange(editorState);
  }

  handleKeyCommand(command) {
    const {editorState} = this.props;
    const newState = RichUtils.handleKeyCommand(editorState, command);
    if (newState) {
      this.props.onChange(newState);
      return true;
    }
    return false;
  }

  setReadOnly(readOnly) {
    this.setState({readOnly});
  }

  mediaBlockRenderer(block) {
    if (block.getType() === "atomic") {
      return {
        component: Media,
        editable: false,
        props: {
          plugins: this.plugins,
          onChange: ::this.onChange,
          editorState: this.props.editorState,
          setReadOnly: this.setReadOnly
        }
      };
    }

    return null;
  }

  blockStyleFn(contentBlock) {
    const type = contentBlock.getType();
    if (type === "unstyled") {
      return "paragraph";
    }
  }

  render() {
    const {editorState} = this.props;
    const plugins = this.plugins;

    return (
      <div className="megadraft">
        <div
          className="megadraft-editor"
          id="megadraft-editor"
          ref="editor">
          <Sidebar
            plugins={plugins}
            editorState={editorState}
            onChange={::this.onChange}/>
          <Editor
            readOnly={this.state.readOnly}
            plugins={plugins}
            blockRendererFn={::this.mediaBlockRenderer}
            blockStyleFn={this.blockStyleFn}
            handleKeyCommand={::this.handleKeyCommand}
            editorState={editorState}
            onChange={::this.onChange} />
          <Toolbar
            editor={this.refs.editor}
            editorState={editorState}
            onChange={::this.onChange}
            actions={this.actions}/>
        </div>
      </div>
    );
  }
}

export const DraftJS = Draft;
