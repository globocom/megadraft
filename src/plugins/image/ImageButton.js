/*
 * Copyright (c) 2016, Globo.com (https://github.com/globocom)
 *
 * License: MIT
 */

import React, {Component} from "react";
import {Entity, AtomicBlockUtils} from "draft-js";

import icons from "../../icons";


export default class BlockButton extends Component {
  onClick(e) {
    e.preventDefault();
    const src = window.prompt("Enter a URL");
    if (!src) {
      return;
    }

    const entityKey = Entity.create("image", "IMMUTABLE", {src});

    this.props.onChange(AtomicBlockUtils.insertAtomicBlock(
      this.props.editorState,
      entityKey,
      "üç∫"
    ));
  }

  render() {
    return (
      <button className={this.props.className} onClick={::this.onClick}>
        <icons.ImageIcon className="sidemenu__button__icon" />
      </button>
    );
  }
}
