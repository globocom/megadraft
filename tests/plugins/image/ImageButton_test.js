/*
 * Copyright (c) 2016, Globo.com (https://github.com/globocom)
 *
 * License: MIT
 */

import React from "react";
import { shallow, mount } from "enzyme";

import ImageButton from "../../../src/plugins/image/ImageButton";
import { editorStateFromRaw } from "../../../src/utils";

describe("ImageButton component", () => {
  let testContext;

  beforeEach(() => {
    const INITIAL_CONTENT = {
      entityMap: {},
      blocks: []
    };

    testContext = {
      onChange: jest.fn(),
      editorState: editorStateFromRaw(INITIAL_CONTENT),
      className: "image-button",
      title: "title image button"
    };

    jest.spyOn(window, "prompt").mockImplementation(() => "Enter an URL");

    testContext.wrapper = mount(<ImageButton {...testContext} />);
  });

  // smoke test
  it("it should render", () => {
    shallow(<ImageButton />);
  });

  it("it should render the expected HTML", () => {
    expect(testContext.wrapper.html()).toMatchSnapshot();
  });

  it("it should render the expected className", () => {
    expect(testContext.wrapper.find("button.image-button")).toHaveLength(1);
  });

  it("it should render the expected attribute title", () => {
    expect(testContext.wrapper.find("button").props()["title"]).toBe(
      "title image button"
    );
  });

  it("it should simulate the click button", () => {
    testContext.wrapper.find("button").simulate("click");
    expect(testContext.onChange.mock.calls.length).toEqual(1);
  });
});
