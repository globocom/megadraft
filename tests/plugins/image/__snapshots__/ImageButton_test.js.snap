// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageButton component it should render the expected HTML 1`] = `"<button class=\\"image-button\\" type=\\"button\\" title=\\"title image button\\"><svg class=\\"sidemenu__button__icon\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\"><path d=\\"M18.222 6H5.778C4.8 6 4 6.6 4 7.333v9.334C4 17.4 4.8 18 5.778 18h12.444C19.2 18 20 17.4 20 16.667V7.333C20 6.6 19.2 6 18.222 6zm-4.084 4l-3 4.51L9 11.503 6 16h12l-3.862-6z\\" fill=\\"currentColor\\" fill-rule=\\"evenodd\\"></path></svg></button>"`;
